### DISCUSSION TITLE
How to add additional firebase functions when deploying nuxt to firebase

### DISCUSSION THREAD

**Discussion by hputzek** [nuxt/nuxt#17084] [Questions]
### Environment

- Operating System: `Darwin`
- Node Version:     `v16.13.1`
- Nuxt Version:     `3.0.0-27324955.23397e6`
- Package Manager:  `npm@8.3.0`
- Bundler:          `Vite`
- User Config:      `css`, `meta`, `vite`, `https`, `ssr`
- Runtime Modules:  `-`
- Build Modules:    `-`

### Reproduction

Prerequisites: A nuxt3 project, e.g. https://codesandbox.io/s/github/nuxt/starter/tree/v3-codesandbox

* Run `NITRO_PRESET=firebase npm run build`

* The firebase function for running nuxt is generated in `.output/server`

* In `/firebase.json` the following config is set:
``` json
{
"functions": { "source": ".output/server" },
}
```

Info: `functions.source` cannot be set to multiple paths

### Nuxt exports only a single function which includes the nuxt build:

`/.output/server/index.mjs`:
```javascript
// [...]
const server = functions.https.onRequest(handle);
export { server };
```

### Describe the bug

Not a bug but a question:

**Is there a way to add/include additional firebase functions that are output to `.output/server` when running  `nuxi build`?**

I want to add a custom firebase function which is deployed to firebase in addition to the nuxt-function when running `firebase deploy`.

But the folder `.output/server` which contains the functions to be deployed is autogenerated by `nuxi build`.
This is also true for the `package.json` file in `.output/server`

I know I can also implement middleware & API endpoints in nuxt, but in my case I explicitly want to use an additional firebase function.

### Additional context

This is what I want to add as a firebase function to existing nuxt3 project: https://maxbarry.medium.com/dynamic-on-demand-image-resizing-using-firebase-hosting-and-google-cloud-functions-to-make-a-cheap-d64e8f5805d1

If I forgot to add important info to this Issue, please let me know so that I can add it.

**Comment by danielroe** [comment]
I've moved this to a discussion, as you observed it was `Not a bug but a question`. But we may end up a creating a new feature request out of this.

Is there a reason you don't want to add a Nitro endpoint to handle image resizing?

**Reply by hputzek** [reply]
@danielroe

> Is there a reason you don't want to add a Nitro endpoint to handle image resizing?

 I read [in this article](https://maxbarry.medium.com/dynamic-on-demand-image-resizing-using-firebase-hosting-and-google-cloud-functions-to-make-a-cheap-d64e8f5805d1) that firebase function results are cached automatically.
Furthermore I could just reuse the code from the article without having to rewrite it to nitro syntax (yes, I'm lazy ;-))

-----

Besides of that I found a way to add additional firebase functions while also deploying the nuxtApp.
But this is VERY hacky, so I would appreciate it a lot if anybody got a better idea how to handle it:

* in your firebase.json  set the functions folder:
```javascript
"functions": { "source": "firebase/functions" },
```

* create folder `./firebase/functions` -> This is where to put additional firebase functions to be deployed
* create a package.json in this folder using `npm init`
* create a file `index.mjs`:
```javascript
import { server } from "nuxt-server"
export { server } // additional functions can be imported and added here
```

Now the important part: The existing output from `.output/server` has to be added in `./firebase/functions/package.json` as a dependency.
The module can NOT just be installed as a local module ([using `file:<my/module>` syntax](https://firebase.google.com/docs/functions/handle-dependencies#including_local_nodejs_modules)). 
Because this just creates a symlink and does not copy the dependencies. -> When attempting to upload the folder using `firebase deploy`you will get an error.

Instead we need all dependencies in the `node_modules`-Folder in `./firebase/functions`  - not symlinked but "real" dependencies.

To achieve this I run the following shell script before deploying to firebase:

```sh
# go to function directory
cd ../../.output/server || exit
# set version & name in package.json generated by nuxt (necessary for 'npm pack' to work)
npm pkg set version='1.0.0'
npm pkg set name='nuxt-server'
# go back to firebase functions folder
cd ./../../firebase/functions/ || exit
# pack directory '/.output/server' and install created package
# This is necessary because symlinks - as they are created by npm install for local packages - do not work when uploading to firebase.
npm install "$(npm pack ../../.output/server | tail -1)"
```

Now the folder `./firebase/functions` (which now also includes the nuxt app as a function) deploys successfully.
Its possible to add more firebase functions to be deployed by adding them into `./firebase/functions` and importing them there in `index.mjs`

**Reply by drm-addoptio** [reply]
I also need to be able to write additional functions logic beside the single function for the server. In my opinion, this is a very important feature, that has to be implemented! And it should be very easy to do that. You could just define a place, where we should put the additional functions, which you could than use to give it to firebase along with the server.

**Reply by ramesh533** [reply]
> Is there a reason you don't want to add a Nitro endpoint to handle image resizing?

@danielroe Is there a way to integrate firebase pubsub functions?

**Reply by andrewspy** [reply]
@danielroe The appeal of using firebase in the first place is its ability to integrate the many services that are offered by firebase via cloud functions, reserving the cloud function only for nuxt server doesn't make much sense.

A common use case for firebase cloud function is the ability to trigger functions on firestore changes. Furthermore, most developers would be using typescript for cloud functions, which cannot be easily injected into `index.mjs`

Would it be possible/easier to just use `NITRO_PRESET=node` and link it to `functions:nuxtServer` end-point? What's the difference between the `firebase` and `node` preset?

**Reply by danielroe** [reply]
I think we should aim to support this in Nitro/nitropack, if it's compatible with the concept. I was asking for clarity just to make sure we implement it rightly.

**Reply by dosstx** [reply]
> I also need to be able to write additional functions logic beside the single function for the server. In my opinion, this is a very important feature, that has to be implemented! And it should be very easy to do that. You could just define a place, where we should put the additional functions, which you could than use to give it to firebase along with the server.

not needed. This is what you just need to do: https://github.com/nuxt/framework/discussions/2414#discussioncomment-2758634

**Comment by drm-addoptio** [comment]
I could find an easy workaround, until this may be solved by the nuxt team:
Basically, you have to change the functions configuration in the firebase.json file during different steps of deploying to firebase. The details are as follows:

1. Ensure your firebase.config is in the origin form, concering the funtions part of it, (for your functions that are not part of the nuxt server hosting),  e.g.:
  
```
"functions": {
    "source": "functions",
    "predeploy": [
      "npm --prefix \"$RESOURCE_DIR\" run lint",
      "npm --prefix \"$RESOURCE_DIR\" run build"
    ]
  },
```
2. run cmd: `firebase deploy --only functions,storage,firestore`
3. change the functions part in firebase.json to: `functions": { "source": ".output/server" },`
4. run cmd: `NITRO_PRESET=firebase nuxi build`
5. run crm: `firebase deploy --only functions:server,hosting`
6. reconfigure the functions part in firebase.json to the origin form

Et voila, you will see it works very smootly. It is a bit more work to deploy, but very acceptable as a workaround. At least for me.

**Reply by danielroe** [reply]
Just to add, it's possible to create [custom Nitro presets](https://v3.nuxtjs.org/docs/deployment/presets/custom) where you can add all this logic. You could start with the [current Firebase preset](https://github.com/nuxt/framework/blob/main/packages/nitro/src/presets/firebase.ts). And if you have any improvements, a PR is welcome!

**Reply by bynicodevelop** [reply]
> Just to add, it's possible to create [custom Nitro presets](https://v3.nuxtjs.org/docs/deployment/presets/custom) where you can add all this logic. You could start with the [current Firebase preset](https://github.com/nuxt/framework/blob/main/packages/nitro/src/presets/firebase.ts). And if you have any improvements, a PR is welcome!

Hello,

Has there been a move in this direction?

**Comment by bynicodevelop** [comment]
Hello, I think my question is a continuation of this post. By default, the function has the name "server".

But how, in the presets, is it possible to rename it?

The goal is to avoid having a conflict with another nuxt project function.

In Firebase you can create several hosting spaces on the same project.
 
But if I have a Nuxt1 project which has the cloud function "server" and the 2nd Nuxt2 project for the same name "server".

Will there be a conflict?

**Reply by vinimarcili** [reply]
Same issue here.
The function deploy overwrite my function in another project with the same name.

**Comment by snakeshift** [comment]
I also want to create additional function.
I use puppeteer and it needs huge memory size.

so, it is reasonable to split function and patch memory size only puppeteer function.

**Comment by LaCocoRoco** [comment]
**You can manage Nuxt3 and Functions as Monorepo**
See Firebase Functions [Documentation](https://firebase.google.com/docs/functions/organize-functions)

Warning A: Minimum Firebase CLI v10.7.1 necessary
Warning B: [Firebase Tools Deploy Bug](https://github.com/nuxt/framework/issues/4961)

Example:
```json
{
  "functions": [
    {
      "source": ".output/server",
      "codebase": "nuxt"
    },
    {
      "source": "functions",
      "codebase": "functions"
    }
  ],
  "hosting": [
    {
      "site": "<your_project_id>",
      "public": ".output/public",
      "cleanUrls": true,
      "rewrites": [
        {
          "source": "**",
          "function": "server"
        }
      ]
    }
  ]
}
```

**Reply by dosstx** [reply]
> **You can manage Nuxt3 and Functions as Monorepo** See Firebase Functions [Documentation](https://firebase.google.com/docs/functions/organize-functions)
> 
> Warning A: Minimum Firebase CLI v10.7.1 necessary Warning B: [Firebase Tools Deploy Bug](https://github.com/nuxt/framework/issues/4961)
> 
> Example:
> 
> ```json
> {
>   "functions": [
>     {
>       "source": ".output/server",
>       "codebase": "nuxt"
>     },
>     {
>       "source": "functions",
>       "codebase": "functions"
>     }
>   ],
>   "hosting": [
>     {
>       "site": "<your_project_id>",
>       "public": ".output/public",
>       "cleanUrls": true,
>       "rewrites": [
>         {
>           "source": "**",
>           "function": "server"
>         }
>       ]
>     }
>   ]
> }
> ```

Fantastic! This is what I was looking for. Now I am able to deploy a firebase app via cloud functions and also listen to database triggers via http cloud functions. Thanks!

**Reply by andrewspy** [reply]
Just want to add that to deploy the nuxt build (i.e. firebase build) to firebase, use the following,

```
firebase deploy --only hosting,functions:nuxt:server
```
